### PR TITLE
Improve performance and usability

### PR DESCRIPTION
**Pull request by Byron**
This PR is a major change which includes:

- parameterize all previously hard-coded values with `clap` on the CLI
- use `anyhow` for error handling and remove unwrap/expect
- improve performance by a factor of 3 by initializing `syntect` only once
- configurable input directory which skips files ignored by .git out of the box to make it feasible

I have made these improvements to allow me to use that tool for visualizing `gitoxide` occasionally,
and I fully understand they might not be desirable here. In that case, please feel free to close the
PR or take what you find valuable.

### CODE REVIEW

**Comment by Byron** [comment]
output

The above is what I see with `cargo run --release -- -i ../gitoxide --ignore-files-without-syntax --line-height-pixels 2 --column-width-pixels 150 --open`. Now I just have to figure out what these 'blocky' files are with long lines in them 😁.

**Comment by Byron** [comment]
output

Much better! (`cargo run --release -- -i ../gitoxide --ignore-extension md --ignore-files-without-syntax --line-height-pixels 1 --column-width-pixels 120 --open`

**Comment by Byron** [comment]
output

Sorry for the spam, but I am having fun :D. (`cargo run --release -- --theme "base16-ocean.dark" -i ../gitoxide --ignore-extension md --ignore-files-without-syntax --line-height-pixels 1 --column-width-pixels 120 --open`)

**Comment by sloganking** [comment]
Hey @Byron, This looks like an excellent PR. Thanks for all the hard work! I will probably be a bit busy over the next few days, but I will definitely look at this as I find the time.

In the meantime. I am not very familiar with `clap` yet. I am aware of `anyhow` but not familiar with using it yet. I am going to look into both of these to see how they both work. But in the meantime, any explanation of what they do and what you are using them for here is appreciated.

**Comment by Byron** [comment]
> But in the meantime, any explanation of what they do and what you are using them for here is appreciated.

`clap` is the standard commandline argument parser in the Rust ecosystem, and `anyhow` is the same but for high-level application errors.

Please note that I cared only about my own convenience while using and tuning the tool, which might make this PR a terrible contribution. It's here mainly for reference and I expect it to be closed. That said, if you need changes, please feel free to push them directly into this PR or take what you need. That's my way of saying, once the tools does what I need it to, I am probably done spending time on it as I should really be doing something else 😅. Thanks for your understanding and I hope what you see here is useful in some way or another.

**Comment by Byron** [comment]
output

(created with `cargo run --release -- --open -i ../gitoxide  --line-height-pixels 1 --ignore-extension md --column-width-pixels 120 --fg-pixel-color style-ascii-brightness`).

I think I am done now - this image is modulating the style color with the actual character value, which yields more interesting patterns in places that otherwise are single-colored. It's a platform to experiment with different color modulations I suppose, but this one already makes it look like it could be tiny characters :D.

**Comment by sloganking** [comment]
> 

I like it. The VsCode scrollbar preview, modulated colors like that too. I suspect it's actually based on the characters themselves instead of randomized, but that has been a feature I've been thinking of as a stretch goal for a while.

**Comment by Byron** [comment]
Screen Shot 2022-09-05 at 12 00 09

The linux kernel!

```
 03:22:02 render Image dimensions: 55900 x 31462 (1.8 GB in memory)
 03:33:05 overall done 77971 files in 664.49s (117 files/s)
 03:33:05  render done 35173654 lines in 664.49s (52933 lines/s)
 03:33:05 overall Longest encountered line in chars: 50204
 03:33:05 overall Aspect ratio is 0.001031226382443684 off from target
```

I have not found an image viewer yet, except for MacOS preview, that could view the whole file properly.

**Comment by sloganking** [comment]
@Byron 

> overall done 77971 files in 664.49s (117 files/s)

You really did triple the performance. I've previously played with rendering the Linux kernel and it took about 30 minutes.

I'm impressed you were able to view that image. What image viewer did you use?

**Comment by sloganking** [comment]
I see `Swatinem/rust-cache@v2` in `rust.yml`. Looking it up, it looks like this caches the compiled libraries, and speeds up future  `rust.yml` building and testing. Is that correct, or is it being used for something more?

**Comment by Byron** [comment]
> You really did triple the performance. I've previously played with rendering the Linux kernel and it took about 30 minutes.

And it could be even faster if `par_iter()` of `rayon` would be used to parallelize over files. This would mean each file's pixels is written into a vec containing all the pixels, which is assembled in one thread that aggregates all of them into the final image. That way it should scale linearly with cores. The highlighting is the main consumer of CPU cycles here, and it could probably be another 25% faster or more if it wouldn't allocate, but it's out of scope here and parallelization is the low-hanging fruit here. It's not easy enough to implement for me to do it though, especially since the linux kernel is exceptional.

> I'm impressed you were able to view that image. What image viewer did you use?

MacOS Finder preview can do it, while `Preview` cannot. However, https://github.com/ArturKovacs/emulsion can also do it, even though it blew up in memory after a little zooming and scrolling.

Screen Shot 2022-09-05 at 12 38 48

> I see Swatinem/rust-cache@v2 in rust.yml. Looking it up, it looks like this caches the compiled libraries, and speeds up future rust.yml building and testing. Is that correct, or is it being used for something more?

It's just that, and I add it to all of my projects these days. It just seems to work.

**Comment by Byron** [comment]
In a moment of weakness I played with output formats to see if there are ones that can be streamed to disk, and `tif` is one of them. This seems to make the venerable `Preview` app happier, too, allowing these pictures to be viewed without killing the app.

Here is the Rust project (sans LLVM), just because (`code-vis -i . --line-height-pixels 1 --column-width-pixels 120 -o output.tif`).

Screen Shot 2022-09-05 at 20 57 24

**Comment by Byron** [comment]
I was also thinking about parallelization and/or memory reduction. Memory reduction could be achieved by creating a memory map of output pixels on disk in write mode, which should allow one or more threads to write into it right away, without having to actually keep all output in memory at all times. Holding the image buffer is the number one use of memory in big projects like the linux kernel.

Once writing to a memory map is implemented (which really is just like writing into memory directly, one could also quite easily implement multi-threading. I imagine the main thread to pre-check all input (something it's doing already) and compute the rectangle into which the writing would have to happen. From there the offsets into memory can be computed to write directly into the correct spot in memory, which equals the disk once the pages are flushed.

If that would be implemented, the only limit is the size of files on disk which are all pre-loaded into memory right now. So, on a 16GB system like mine, right now the linux kernel takes 2GB'ish in memory for files alone, but on top of that it's 12.7GB for the frame buffer, and with the changes I'd think it can handle 7 times more so about 14GB in memory, the rest of the 89GB image is written straight to disk :D. Then, one would probably just need to create a viewer application that streams picture data to view it ;).

**Comment by sloganking** [comment]
@Byron It should be noted that `.tiff` files have a hardcoded filesize limit of 4GB. They are also either have no compression or much less compression than ``.png`` for example. Is the Linux kernel image even store-able in a `.tiff`?

**Comment by sloganking** [comment]
@Byron I have already done some work on a large image viewer that can handle displaying a `png` image of any size. The Linux kenel render inspired it. It's not a public project yet but there may be other options if you're having a hard time displaying large images.

Also I think `.psb` files can handle super large images. However it's made by adobe and I'm not sure what support there is for writing to or displaying them. I don't think the [image crate](https://crates.io/crates/image) supports them at the moment.

For your convenience, Nasa has a 69536 x 22230 px, 4.3GB `.psb` image [here](https://esahubble.org/images/heic1502a/)

**Comment by alpha-tango-kilo** [comment]
> I was also thinking about parallelization and/or memory reduction. Memory reduction could be achieved by creating a memory map of output pixels on disk in write mode, which should allow one or more threads to write into it right away, without having to actually keep all output in memory at all times. Holding the image buffer is the number one use of memory in big projects like the linux kernel.
> 
> Once writing to a memory map is implemented (which really is just like writing into memory directly, one could also quite easily implement multi-threading. I imagine the main thread to pre-check all input (something it's doing already) and compute the rectangle into which the writing would have to happen. From there the offsets into memory can be computed to write directly into the correct spot in memory, which equals the disk once the pages are flushed.

I've toyed with mmaps before for other applications, would be happy to take a look at this either on your fork or after this PR gets merged. The only challenge is if you wanted to explicitly enforce a ceiling on memory usage during runtime - then you have a whole other problem of deciding what amount to choose and how to keep to it - however just relying on the throughput-then-drop of mmaps should definitely yield benefits over the current approach, even if it's not as memory-minimal as it could be.

**Comment by Byron** [comment]
> @Byron I have already done some work on a large image viewer that can handle displaying a png image of any size. The Linux kenel render inspired it.

That's great, I am looking forward to giving it a whirl once it's available.

> I don't think the [image crate](https://crates.io/crates/image) supports them at the moment.

> The only challenge is if you wanted to explicitly enforce a ceiling on memory usage during runtime - then you have a whole other problem of deciding what amount to choose and how to keep to it - however just relying on the throughput-then-drop of mmaps should definitely yield benefits over the current approach, even if it's not as memory-minimal as it could be.

Ceiling memory could be controlled by letting the user decide when to give up, and 'guessing' the amount of memory used to get a ballpark number should be easy enough. In any case, having disk-backed writable memory maps should allow going beyond what the tool can do now  and pave the way for parallel processing as well. The image format question would resolve itself as it had to be one that ideally is an uncompressed format that has no size limit. Alternatively, one could use the memory map as temporary storage and save the image later using traditional means.

> I've toyed with mmaps before for other applications, would be happy to take a look at this (of a weekend) either on your fork or after this PR gets merged.

It's hard for me to resist the urge, it might be done by then 😅.

**Comment by Byron** [comment]
With anonymous mmaps it's already possible to generate the linux kernel with 120 characters per line, up from 50, and it could probably handle more.  It's beautiful to see the virtual memory system kick in, compressing memory on the fly and generally having no issue with the load, despite needing 12GB just for the image buffer and with all other apps open. Saving the image as png also takes a couple of additional (real) memory gigabytes as it compresses the image buffer into memory first, while other formats can flush the buffer to disk more directly (but don't support these dimensions of 86k*48k :D).

Here is the log…:

```
❯ code-vis -i .   --column-width-pixels 120 --line-height-pixels 1
 render Image dimensions: 86640 x 48717 x 3 (12.7 GB in memory)
 overall done 77971 files in 678.49s (114 files/s)
  render done 35173654 lines in 678.49s (51841 lines/s)
 overall Longest encountered line in chars: 50204
 overall Aspect ratio is 0.0006568548966481291 off from target
  saving done 2.0GB in 69.18s (29.1MB/s)
```

…and here is the output (the quickview app is able to view the image in screen resolution, but that's all):

Screen Shot 2022-09-06 at 08 59 55

File: .gitignore [hunk 1]
```diff
@@ -10,9 +10,9 @@ Cargo.lock
 **/*.rs.bk
 
 
-input/
```
Comment (sloganking) [inline .gitignore:13]: Why were these modified?

Comment (Byron) [reply .gitignore:13]: To only match the top-level `input` directory, not the ones deeper down in the hierarchy.

**Comment by sloganking** [comment]
Does anyone know why I can't compile this with the `CodeLLDB` debugger in vscode?

```
Running `cargo build --bin=code-vis --package=code-visualizer --message-format=json`...
   Compiling anyhow v1.0.64
   Compiling globset v0.4.9
   Compiling crossterm v0.22.1
   Compiling syntect v4.6.0
   Compiling rayon v1.5.3
   Compiling clap v3.2.20
error[E0407]: method `provide` is not a member of trait `StdError`
   --> /home/fire/.cargo/registry/src/github.com-1ecc6299db9ec823/anyhow-1.0.64/src/context.rs:131:5
    |
131 | /     fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
132 | |         self.error.provide(demand);
133 | |     }
    | |_____^ not a member of trait `StdError`
```

**Comment by Byron** [comment]
Oh my, it's done! Here is the linux kernel after using all cores (but it's not a benchmark, was doing other things as well), shrinking more than 11 minutes for rendering down to 1m 40s. Storing a png file takes a whopping 68s for a total runtime of close to 3 minutes.

```
❯ code-vis -i .   --column-width-pixels 120 --line-height-pixels 1
 render Image dimensions: 86640 x 48717 x 3 (12.7 GB in virtual memory)
 render Aspect ratio is 0.0006568548966481291 off from target
 overall done 77971 files in 101.59s (767 files/s)
  render done 35173654 lines in 101.59s (346240 lines/s)
 overall Longest encountered line in chars: 50204
  saving done 2.0GB in 67.59s (29.8MB/s)
 operation done in 169.55s
```

Threaded mode is the default now and `gitoxide` is done in 0.44s. My next goal is to make it take 1s again (by writing more code :D).

Note that one could possibly tune a few things here and there to get a few percentage points, but I didn't feel the need to even look.

**Comment by Byron** [comment]
And as bonus, now that there is no resistance anymore to visualize even the biggest projects, here is LLVM…
llvm

…and Rust:

rust

**Comment by Byron** [comment]
That was unexpected: I thought I was done and just wanted to visualize WebKit… and then it hangs, always. Turns out that `syntect` can encounter lines that show exponential behaviour when being highlighted, or maybe even get into a loop. Initially I thought it's due to lines being too long, so I ran the highlighting on truncated lines. That was a bad idea as the linux kernel now also failed to render. Thinking about it, it made sense that the highlighter being state dependent, providing incorrect syntax is probably more likely to trigger bugs. After that performance feature was hidden behind a flag and disabled, another flag was added to print the line that is about to be processed. With that it was possible to see that `regex-dna.js` kill `syntect` and `--ignore-extension js` would make it possible to see the first version of WebKit.

webkit

```
❯ code-vis -i . --column-width-pixels 120 --line-height-pixels 1 --ignore-extension js
 search unicode files done 387907 files in 85.12s (4557 files/s)
              content done 1.2GB in 85.12s (14.2MB/s)
                input Ignored 72190 files that matched ignored extensions
               render Image dimensions: 73320 x 41294 x 3 (9.1 GB in virtual memory)
               render Aspect ratio is 0.0022171636449739296 off from target
              process done 271445 files in 65.37s (4152 files/s)
               render done 25230198 lines in 65.37s (385986 lines/s)
              process Longest encountered line in chars: 6646057
               saving done 2.0GB in 108.32s (18.3MB/s)
            operation done in 174.01s
```
(_it takes longer to save the image than the rendering itself_)

Good enough, I think one could consider filing an issue against `syntect` to see if 'regex-dna.js' can be made to work.

File: src/render.rs [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+
+            (line_num, longest_line_chars, background)
+        } else {
+            let mut line_num: u32 = 0;
+            let mut longest_line_chars = 0;
+            let mut background = None;
+            std::thread::scope(|scope| -> anyhow::Result<()> {
+                let (tx, rx) = flume::bounded::<(_, String, _, _)>(content.len());
+                let (ttx, trx) = flume::unbounded();
+                for tid in 1..threads {
```
Comment (sloganking) [inline src/render.rs:313]: Shouldn't this be `0..threads`?

File: src/main.rs [hunk 3]
```diff
@@ -1,11 +1,123 @@
...
+            .checked_duration_since(start)
+            .unwrap_or_default()
+            .as_secs_f32()
+    ));
+
+    render_progress.shutdown_and_wait();
+    Ok(())
+}
+
+fn sage_image(
```
Comment (sloganking) [inline src/main.rs:97]: I assume this is supposed to be spelled `save_image()`

**Comment by sloganking** [comment]
I think I have a pretty good high level understanding of the changes so far. There might be a few small details that elude me, and there could be a little more documentation, but I am happy with how this is for now. I am going to merge and clean from there.

**Comment by sloganking** [comment]
Thanks for all the help @Byron. This looks great!

File: src/render.rs (hunk 2 above)
Comment (Byron) [reply src/render.rs:313]: It does count the main thread as working thread as it aggregates, thus it excludes it. It's probably OK to overcommit a little but I wanted to be sure the aggregator will not be starved as it will slow down all producers.

File: src/main.rs (hunk 3 above)
Comment (Byron) [reply src/main.rs:97]: Most definitely, even though the image looks very magical to me sometimes :D.

**Review by sloganking — COMMENTED** [review]
Looks good
